core/src/main/java/org/apache/calcite/adapter/enumerable/EnumUtils.java

```diff
@@ -796,4 +803,244 @@ static Expression windowSelector(
+              finalWindowElementsMap.put(currentWindow, new ArrayList<>(tempElementList));
+            }
+
+            currentWindow = sessionEntry.getKey();
+            tempElementList.clear();
+            tempElementList.addAll(sessionEntry.getValue());
+          } else {
+            // merge windows.
+            currentWindow = mergeWindows(currentWindow, sessionEntry.getKey());
+            // merge elements in windows.
+            tempElementList.addAll(sessionEntry.getValue());
+          }
+        }
+
+        if (!tempElementList.isEmpty()) {
+          finalWindowElementsMap.put(currentWindow, new ArrayList<>(tempElementList));
+        }
+
+        // construct final results from finalWindowElementsMap.
+        for (Map.Entry<Pair<Long, Long>, List<Object[]>> finalWindowElementsEntry
+            : finalWindowElementsMap.entrySet()) {
+          for (Object[] element : finalWindowElementsEntry.getValue()) {
+            Object[] curWithWindow = new Object[element.length + 2];
+            System.arraycopy(element, 0, curWithWindow, 0, element.length);
+            curWithWindow[element.length] = finalWindowElementsEntry.getKey().left;
+            curWithWindow[element.length + 1] = finalWindowElementsEntry.getKey().right;
+            list.offer(curWithWindow);
+          }
+        }
+      }
+    }
+
+    private boolean isOverlapped(Pair<Long, Long> a, Pair<Long, Long> b) {
+      return !(b.left >= a.right);
+    }
+
+    private Pair<Long, Long> mergeWindows(Pair<Long, Long> a, Pair<Long, Long> b) {
+      return new Pair<>(a.left <= b.left ? a.left : b.left, a.right >= b.right ? a.right : b.right);
+    }
+
+    private Pair<Long, Long> computeInitWindow(long ts, long gap) {
+      return new Pair<>(ts, ts + gap);
+    }
+  }
+
+  /**
+   * Create enumerable implementation that applies hopping on each element from the input
+   * enumerator and produces at least one element for each input element.
+   */
+  public static Enumerable<Object[]> hopping(Enumerator<Object[]> inputEnumerator,
+      int indexOfWatermarkedColumn, long emitFrequency, long intervalSize) {
+    return new AbstractEnumerable<Object[]>() {
+      @Override public Enumerator<Object[]> enumerator() {
+        return new HopEnumerator(inputEnumerator,
+            indexOfWatermarkedColumn, emitFrequency, intervalSize);
+      }
+    };
+  }
+
+  private static class HopEnumerator implements Enumerator<Object[]> {
+    private final Enumerator<Object[]> inputEnumerator;
+    private final int indexOfWatermarkedColumn;
+    private final long emitFrequency;
+    private final long intervalSize;
+    private LinkedList<Object[]> list;
+
+    HopEnumerator(Enumerator<Object[]> inputEnumerator,
+        int indexOfWatermarkedColumn, long emitFrequency, long intervalSize) {
+      this.inputEnumerator = inputEnumerator;
+      this.indexOfWatermarkedColumn = indexOfWatermarkedColumn;
+      this.emitFrequency = emitFrequency;
```

Comment:
`indexOfWatermarkedColumn ` -> `wmIndex` ?

Reply:
I prefer to keep "watermark". This is the relatively new concept for Calcite, so abbreviation could make people confused. 

Actually I have observed it on "TVF" (table-valued function or table function). Seems like people are already confused on such abbreviation.

Reply:
Use "table function" sounds better, i don't really know what a "table-valued function" means.